### ISSUE TITLE
Questions on GANSynth training

### ISSUE THREAD

**Issue by keunwoochoi** [tensorflow/magenta#1524]
Hi, thanks for the great work. I have some questions about GANSynth training from the paper. I figured out some of them from the codes but couldn't find everything. (cc' @jesseengel @chrisdonahue) 

- Q 1-1. From Section D, On concatenating (x, minibatch std), I guess it means - if the `x=` the tensor size of `(batch_size, 2, 16, 256)`, the since the `std` channel size is `(None, 2, 16, 1)`, the std is computed for each pixel & over channel & over mini batch items? I.e., `x.std(axis=0, keepdims=True).std(axis=3, keepdims=True)`? 
- Q 1-2. Could you tell me why did we need this?

I'm also confused by (Section D)
> measure the maximum range over 100 examples and independently shift and scale the log-magnitudes and phases to [-0.8, 0.8]

, especially by 'measure.. over 100 examples' vs 'independently'. Also, unlike section D in the paper, it says in Section 2.2 that 
> ..then scale the magnitudes to be between -1 and 1 to..  ..the phase angle is also scaled to between -1 and 1 .. as **phase** models. 

Also later,
> ..unwrap the phase angle and take the finite difference .. **IF**

To summarize, 
- Q 2-1.  What is happening to log magnitude spectrograms? is it [-0.8, 0.8] or [-1, 1] (independently? how are the information from 100 examples are used?)
- Q 2-2. What is happening to the phases in Phase model?
- Q 2-3. Same for IF model?

I would very appreciate some concrete explanation on it. Thanks!

**Comment by chrisdonahue** [comment]
Hi Keunwoo! Thanks for your interest in the work (and for defending it on twitter :). I will try to answer your questions below, though other authors may know more.

I am not sure about the answers to Q 1-1 and Q 1-2. I think Jesse or Shuo would be the right people to ask about that as they were more closely involved with the design of the architecture.

For Q 2-1, we find the minimum (a) and maximum (b) values of 100 (somewhat arbitrary choice) log-amplitude spectrograms from the training set. We normalize all of the dataset spectrograms by a linear rescaling from [a, b] -> [-0.8, 0.8]. Because a and b were computed from only 100 examples, other examples in the dataset might have values outside of [-0.8, 0.8] after rescaling, but likely not by much considering that they are log amplitudes. I am not sure what is meant by the term "independently" (I didn't write this section), but one of the other authors might know.

For Q 2-2, the phase is scaled from [-pi, pi] to [-1, 1]. I am not sure why we didn't also scale these values to [-0.8, 0.8], given that the generator also has a tanh nonlinearity for producing phase.

For Q 2-3, Instaneous frequency is computed by taking the finite differences across the time axis of the unwrapped phase spectra. The range of IF values is thus [-2pi, 2pi]. I assume (but am not certain) that we are rescaling this to [-1, 1].

If another author could chime in to verify my explanations, that would be helpful :)

**Comment by keunwoochoi** [comment]
Hi Chris! Thanks for your answer. Especially..
> :)

Was it possible to close a parenthesis with :)? Whoa 🤯

> "independently"

Then it probably means the processing has been done for each item --  'data sample-wisely'.

**Comment by chrisdonahue** [comment]
Haha yeah I saw someone close their parentheses with a smiley face once and adopted the practice.

Yes I believe this means it is done for each item.

**Comment by keunwoochoi** [comment]
> the finite differences across the time axis of the unwrapped phase spectra. The range of IF values is thus [-2pi, 2pi]

I think it could depend on how phase unwrapping was implemented. Was testing with numpy's unwrap function, and the finite difference is precisely in a range of `[-pi, pi]`.

I'll keep it open for some while (i.e. forever) since there's still little ambiguity, although most of my questions were answered :)

**Comment by chrisdonahue** [comment]
Ah right. It depends on what you set as the maximum discontinuity values for `np.unwrap`. But with the default configuration of `np.pi` (which we also use), you are correct that the range should be [-pi, pi]. If you use a larger discontinuity `> np.pi`, the range will increase accordingly.

**Comment by jesseengel** [comment]
Also, fwiw, I followed np.unwrap closely when writing the tensorflow
version.

*From: *Chris Donahue <notifications@github.com>
*Date: *Mon, May 13, 2019 at 8:46 PM
*To: *tensorflow/magenta
*Cc: *Jesse Engel, Mention

Ah right. It depends on what you set as the maximum discontinuity values

**Comment by ss12f32v** [comment]
Hi, there is one additional question about GANsynth training here.
Did the author meet the mode collapse scenario during the training?
It seems that sometimes the model training process is not stable, although the loss is using WGAN-GP.

**Comment by jesseengel** [comment]
Hi,

In our experience the training is quite stable (both for progressive GAN
and not progressive), but must be trained for the full number of
iterations. Early in training, the GAN has only learned a few modes (which
can look like mode collapse), but by the end of training there is a lot
more diversity.

Best,
Jesse
